### PR TITLE
Improve security

### PR DESCRIPTION
**Pull request by nplasterer**
- prevent injecting messages across topics i.e. rejecting any messages that weren't encrypted/signed for this specific topic.
- prevent an attacker (or compromised XMTP server) from sneaking an invalid/spoofed contact bundle into the Contact topic.
- clean up getting public key from private key

### CODE REVIEW

**Comment by nplasterer** [comment]
@nakajima some of these fixes might need to happen in iOS as well. FYI

File: library/src/main/java/org/xmtp/android/library/Contacts.kt [hunk 1]
```diff
@@ -29,6 +30,13 @@ data class Contacts(
         for (envelope in response.envelopesList) {
             knownBundles[peerAddress] = ContactBundleBuilder.buildFromEnvelope(envelope)
```
Comment (michaelx11) [inline library/src/main/java/org/xmtp/android/library/Contacts.kt:31]: We can filter in the for loop, so the code can tolerate a few bad bundles and still return the first legit one.

File: library/src/main/java/org/xmtp/android/library/messages/MessageV2.kt [hunk 2]
```diff
@@ -64,10 +64,13 @@ class MessageV2Builder {
             }.build()
 
             if (key.walletAddress != (PublicKeyBuilder.buildFromSignedPublicKey(signed.sender.preKey).walletAddress)) {
-                throw throw XMTPException("Invalid signature")
+                throw XMTPException("Invalid signature")
             }
             val encodedMessage = EncodedContent.parseFrom(signed.payload)
             val header = MessageHeaderV2.parseFrom(message.headerBytes)
+            if (header.topic != topic) {
```
Comment (michaelx11) [inline library/src/main/java/org/xmtp/android/library/messages/MessageV2.kt:71]: nice!

**Review by michaelx11 — APPROVED** [review]
Thanks for the quick changes! One requested change on contact bundle, otherwise LGTM

File: library/src/main/java/org/xmtp/android/library/Contacts.kt (hunk 1 above)
Comment (nplasterer) [reply library/src/main/java/org/xmtp/android/library/Contacts.kt]: Ahh good call. I fixed that in https://github.com/xmtp/xmtp-android/pull/53/commits/b6ead02697418abcddf625a4f1da55b2e8e8573c

**Review by nakajima — APPROVED** [review]

**Pull request merged**